Added option to make included Taskfile run commands on its own directory

evg4b opened the pull request:
Hi, this pull request relates to [#260](https://github.com/go-task/task/issues/260) and [#144](https://github.com/go-task/task/issues/144) issues. This syntax was implemented:

``` yaml
includes:
  docs:
    taskfile: ./docs/Taskfile.yml
    dir: ./docs
```

Also added support `Taskfile_{GOOS}.yaml` for included task files.

andreynering commented on internal/taskfile/included_taskfile.go:
```diff
@@ -0,0 +1,38 @@
+
+var (
+	// ErrCantUnmarshalIncludedTaskFile is returned for invalid var YAML.
+	ErrCantUnmarshalIncludedTaskFile = errors.New("task: can't unmarshal included value")
```
`TaskFile` -> `Taskfile`

andreynering commented on internal/taskfile/included_taskfile.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// IncludedTaskFile represents information about included tasksfile
+type IncludedTaskFile struct {
```
`IncludedTaskFile` -> `IncludedTaskfile`

andreynering commented on internal/taskfile/included_taskfile.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+// IncludedTaskFiles represents information about included tasksfiles
+type IncludedTaskFiles = map[string]IncludedTaskFile
```
`IncludedTaskFiles` -> `IncludedTaskfiles`

andreynering commented on internal/taskfile/read/taskfile.go:
```diff
@@ -44,6 +45,24 @@ func Taskfile(dir string, entrypoint string) (*taskfile.Taskfile, error) {
+
+		includedTaskDirectory := filepath.Dir(path)
+		includedTaskfileName := strings.TrimSuffix(filepath.Base(path), filepath.Ext(path))
+		path = filepath.Join(includedTaskDirectory, fmt.Sprintf("%s_%s.yml", includedTaskfileName, runtime.GOOS))
```
Hey @evg4b,

Actually, including these files automatically going in the reverse direction I had in mind.

I planned to deprecate the automatic inclusion of Taskfiles by OS in v3, and support templating to something like this would work:

```yaml
includes:
  taskfile: Taskfile_{{OS}}.yml
  namespace: false
```

(The `namespace: false` part needs thought, but we would need a way allow overwriting instead of namespacing).

I propose that you just remove the automatic inclusion for this PR. No need to implement the above now, I can do that in another PR.

What do you think?

evg4b replied:
Sounds good.
Sorry, I did not carefully read the roadmap. I removed this functionality.

andreynering commented on internal/taskfile/read/taskfile.go:
```diff
@@ -44,6 +45,24 @@ func Taskfile(dir string, entrypoint string) (*taskfile.Taskfile, error) {
+		}
+
+		for _, task := range includedTaskfile.Tasks {
+			task.Dir = filepath.Join(includedTask.Dir, task.Dir)
```
We need to skip joining if the path in the task is absolute:

```go
if !filepath.IsAbs(task.Dir) {
	task.Dir = filepath.Join(includedTask.Dir, task.Dir)
}
```

evg4b replied:
You're right, also I fixed this point for included files (`internal\taskfile\read\taskfile.go:32`), it works incorrectly in some cases.
Example:
![image](https://user-images.githubusercontent.com/10252305/74590302-39f7bf00-501e-11ea-9467-0418626c1159.png)

andreynering commented on testdata/includes/module1/Taskfile.yml:
```diff
@@ -0,0 +1,9 @@
+  gen_dir:
+    cmds:
+      - echo included_directory_without_dir > included_directory_without_dir.txt
+  gen_file:
```
Add an empty line to separate tasks

andreynering commented on testdata/includes/module2/Taskfile.yml:
```diff
@@ -0,0 +1,9 @@
+  gen_dir:
+    cmds:
+      - echo included_directory_with_dir > included_directory_with_dir.txt
+  gen_file:
```
Add an empty line to separate tasks

andreynering commented on testdata/includes/module1/Taskfile.yml:
```diff
@@ -0,0 +1,9 @@
+version: '2'
```
Those test files should all be in version 3. Also, we should disallow setting a dir if version is 2. Just returning an error if version is 2 and dir is not `""` is fine. This should be on `Setup()` like the other version validations.

andreynering commented:
If you don't mind, adding an entry in the CHANGELOG and updating documentation would also be nice.

If your time is short, I can do that once I merge this.

evg4b commented:
> If you don't mind, adding an entry in the CHANGELOG and updating documentation would also be nice.
> 
> If your time is short, I can do that once I merge this.

I fixed all your comments and added an entry in the CHANGELOG.md, but updating documentation can be difficult for me because my English isn't good.

andreynering commented:
@evg4b Seems great, thanks!

I'll add the documentation myself, no problem.
